### PR TITLE
feat(reachability): add join cookie

### PR DESCRIPTION
**Pull request by torgeadelin**
# COMPLETES [#SPARK-382449](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-382449)

## This pull request addresses
- grab joinCookie from reachability request
- store joinCookie in localStorage 
- grab join Cookie from localStorage and add it to join call methods
 - `joinMeeting`
 - `joinMeetingWithRoap`
 - `sendRoap`
- small refactoring

Reference: https://confluence-eng-gpk2.cisco.com/conf/pages/viewpage.action?spaceKey=CLO&title=Client+Changes#ClientChanges-ObservedClientPublicMediaIP

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios where tested

- unit tests added for request and other functions

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 1]
```diff
@@ -49,23 +66,24 @@ export default class RoapRequest extends StatelessWebexPlugin {
...
         localMedias: [
           {
-            localSdp: JSON.stringify(
-              this.attachRechabilityData({
-                roapMessage: options.roapMessage,
-                audioMuted: false,
-                videoMuted: false,
-              })
-            ),
+            localSdp: JSON.stringify(reachabilityData.localSdp),
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/roap/request.ts:81]: `stringify` here seems like a bug to me. 
Is it?

Same for `sendRoap`

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 2]
```diff
@@ -49,23 +66,24 @@ export default class RoapRequest extends StatelessWebexPlugin {
       const deviceUrl = this.webex.internal.device.url;
       let url = '';
 
+      const reachabilityData = this.attachReachabilityData({
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/roap/request.ts:69]: BTW, it looks like joinMeetingWithRoap() is not used anywhere and it's removed on beta, so not sure if you need to do this or just delete joinMeetingWithRoap()

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 3]
```diff
@@ -154,36 +179,29 @@ export default class RoapRequest extends StatelessWebexPlugin {
     Metrics.postEvent({event: eventType.MEDIA_REQUEST, meetingId});
 
     // @ts-ignore
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/roap/request.ts:181]: is ths ts-ignore still needed if we don't call this.webex anymore?

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/request.js [hunk 4]
```diff
@@ -190,6 +203,30 @@ describe('plugin-meetings', () => {
         assert.equal(requestParams.uri, 'locusUrl/loci/call?alternateRedirect=true');
         assert.equal(requestParams.body.invitee.address, 'sipUrl');
       });
+
+      it.only('includes joinCookie correctly', async () => {
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/request.js:207]: .only

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 5]
```diff
@@ -143,6 +160,14 @@ export default class RoapRequest extends StatelessWebexPlugin {
       LoggerProxy.logger.info('Roap:request#sendRoap --> Race Condition /call mediaID not present');
     }
 
+    const reachabilityData = this.attachReachabilityData({
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/src/roap/request.ts:163]: the name `reachabilityData` is not great, becuase this is in fact an SDP with audio/video mute information and reachability data, not sure what a better name would be though... maybe localSdpWithReachabilityData?

File: packages/@webex/plugin-meetings/src/meeting/request.ts [hunk 6]
```diff
@@ -32,13 +33,32 @@ import {Reaction} from '../reactions/reactions.type';
...
+   * @param {Object} attrs
+   * @param {Object} options
+   */
   constructor(attrs: any, options: any) {
     super(attrs, options);
     this.changeVideoLayoutDebounced = debounce(this.changeVideoLayout, 2000, {
       leading: true,
       trailing: true,
     });
+    const joinCookieRaw = window.localStorage.getItem(REACHABILITY.localStorageJoinCookie);
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/request.ts:51]: refactored

File: packages/@webex/plugin-meetings/src/roap/request.ts [hunk 7]
```diff
@@ -89,12 +107,11 @@ export default class RoapRequest extends StatelessWebexPlugin {
       }
 
       // @ts-ignore
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/roap/request.ts]: yeah, it's not smart enough...

File: packages/@webex/plugin-meetings/src/roap/request.ts (hunk 2 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/roap/request.ts]: removed (PR based off beta now)

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/request.js (hunk 4 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/request.js]: removed

File: packages/@webex/plugin-meetings/src/roap/request.ts (hunk 5 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/roap/request.ts:88]: refactored

File: packages/@webex/plugin-meetings/test/unit/spec/roap/request.ts [hunk 8]
```diff
@@ -1,114 +0,0 @@
-import {assert} from '@webex/test-helper-chai';
```
Comment (marcin-bazyl) [inline packages/@webex/plugin-meetings/test/unit/spec/roap/request.ts:1]: why has this test file been renamed from ts to js?

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/test/unit/spec/roap/request.ts:1]: i think when this was pointing to master and didn't have the beta changes in, it was js and when i pulled the changes from  beta and point it to beta, i ended up with 2 files, i didn't notice i deleted the ts file.

changed back to ts

**Review by marcin-bazyl — APPROVED** [review]

**Pull request merged**